### PR TITLE
Consume images from Azure Shared Image Gallery

### PR DESCRIPTION
**Pull request by mbearup**
Enable packer to consume images from Azure Shared Image Gallery. 
WIP = still working on tests. The rest of the code should be ready for review.

### CODE REVIEW

File: builder/azure/arm/config.go [hunk 1]
```diff
@@ -79,6 +79,13 @@ type Config struct {
 	CaptureNamePrefix    string `mapstructure:"capture_name_prefix"`
 	CaptureContainerName string `mapstructure:"capture_container_name"`
 
+	// Shared Gallery
+	SharedGallerySubscription  string `mapstructure:"shared_gallery_subscription"`
```
Comment (boumenot) [inline builder/azure/arm/config.go]: Please follow the same pattern as was used for plan_info.  Basically, add these elements to a struct.

File: builder/azure/arm/config.go [hunk 2]
```diff
@@ -572,19 +579,36 @@ func assertRequiredParametersSet(c *Config, errs *packer.MultiError) {
...
+		if c.SharedGallerySubscription == "" {
+			errs = packer.MultiErrorAppend(errs, fmt.Errorf("A shared_gallery_subscription must be specified"))
+		}
+		if c.SharedGalleryResourceGroup == "" {
+			errs = packer.MultiErrorAppend(errs, fmt.Errorf("A shared_gallery_resource_group must be specified"))
+		}
+		if c.SharedGalleryImageName == "" {
+			errs = packer.MultiErrorAppend(errs, fmt.Errorf("A shared_gallery_image_name must be specified"))
+		}
+        if c.CaptureContainerName != "" {
```
Comment (boumenot) [inline builder/azure/arm/config.go]: The `if` spacing seems off.  Make sure to run through gofmt.

File: builder/azure/arm/template_factory.go [hunk 3]
```diff
@@ -72,6 +72,18 @@ func GetVirtualMachineDeployment(config *Config) (*resources.Deployment, error)
 			config.ImageVersion)
 
 		builder.SetManagedMarketplaceImage(config.Location, config.ImagePublisher, config.ImageOffer, config.ImageSku, config.ImageVersion, imageID, config.managedImageStorageAccountType)
+	} else if config.SharedGallerySubscription != "" && config.SharedGalleryResourceGroup != "" && config.SharedGalleryName != "" && config.SharedGalleryImageName != "" {
```
Comment (boumenot) [inline builder/azure/arm/template_factory.go]: Do you need all of these checks, or is it sufficient to use just SharedGallerySubscription?  If you've gotten this far the configuration checks had to all pass.

File: builder/azure/common/template/template_builder.go [hunk 4]
```diff
@@ -145,6 +145,21 @@ func (s *TemplateBuilder) SetManagedMarketplaceImage(location, publisher, offer,
 	return nil
 }
 
+func (s *TemplateBuilder) SetSharedGalleryImage(location, imageID string) error {
+	resource, err := s.getResourceByType(resourceVirtualMachine)
+	if err != nil {
+		return err
+	}
+
+	s.setVariable("apiVersion", "2018-04-01") // Required for Shared Image Gallery
```
Comment (boumenot) [inline builder/azure/common/template/template_builder.go]: Please implement a test that asserts the template is correct when using a shared image.

File: builder/azure/arm/config.go (hunk 2 above)
Comment (mbearup) [reply builder/azure/arm/config.go]: Yeah I ran it through gofmt but it didn't catch this. It's a space vs tab issue...

File: builder/azure/common/template/template_builder.go (hunk 4 above)
Comment (mbearup) [reply builder/azure/common/template/template_builder.go]: Done

File: builder/azure/common/template/template_builder_test.go [hunk 5]
```diff
@@ -181,3 +181,42 @@ func TestBuildWindows02(t *testing.T) {
...
+	err = testSubject.SetSharedGalleryImage("westcentralus", imageID)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	if (*testSubject.template.Variables)["apiVersion"] != "2018-04-01" {
+		t.Fatal("ARM template for Shared Image Gallery must use apiVersion 2018-04-01")
+	}
+
+	foundImageID := false
```
Comment (boumenot) [inline builder/azure/common/template/template_builder_test.go]: I was expecting more a test like TestBuildLinux02, which uses the approval library.  You create a JSON document using the NewTemplateBuilder, and serialize it to disk. The approval library verifies that the output matches the golden copy that is checked into source control.

**Pull request merged**